### PR TITLE
Add initial VLSI flow scripts

### PR DESCRIPTION
**Pull request by jerryz123**
This is not usable, but this is all the non-tech specific stuff needed to run the VLSI flow. Tech-specific repos should branch off from here.

### CODE REVIEW

File: vlsi/example-vlsi [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+def example_place_tap_cells(x: hammer_vlsi.HammerTool) -> bool:
+    x.append('''
+# TODO
+''')
+    return True
+
+class ExampleDriver(CLIDriver):
+
+    def action_map(self) -> Dict[str, Callable[[hammer_vlsi.HammerDriver, Callable[[str], None]], Optional[dict]]]:
```
Comment (colinschmidt) [inline vlsi/example-vlsi]: This isn't the right way to simply add a step.

File: vlsi/Makefile [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+# import other necessary rules and variables
+#########################################################################################
+include $(base_dir)/common.mk
+
+#########################################################################################
+# vlsi types and rules
+#########################################################################################
+
+#sim_name is unused, but GenerateSimFiles expects it
+sim_name     ?= vcs
```
Comment (colinschmidt) [inline vlsi/Makefile]: What is this set for?

File: vlsi/Makefile [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+tech_name    ?= 
+tech_dir     ?= $(vlsi_dir)/hammer-$(tech_nam)-plugin/$(tech_name)
```
Comment (colinschmidt) [inline vlsi/Makefile]: typo, `tech_nam`

File: vlsi/example-vlsi (hunk 1 above)
Comment (jerryz123) [reply vlsi/example-vlsi]: This is all taken from @jwright6323 's intech22 branch of the old project-template

Comment (jwright6323) [reply vlsi/example-vlsi]: 🔥

File: vlsi/Makefile (hunk 2 above)
Comment (jerryz123) [reply vlsi/Makefile]: https://github.com/ucb-bar/project-template/blob/9622a9757067e9c1dab9a4baf551ce932dc87b54/common.mk#L34-L35

File: vlsi/example-vlsi (hunk 1 above)
Comment (colinschmidt) [reply vlsi/example-vlsi]: You just overload `get_extra_par_hooks`
```
def get_extra_par_hooks(self) -> List[HammerToolHookAction]:
  return [hammer_vlsi.HammerTool.make_replacement_hook("place_tap_cells", example_place_tap_cells)]
```

**Comment by colinschmidt** [comment]
Please add a `setup-vlsi.sh` that sources the hammer sourceme and such.

**Comment by jerryz123** [comment]
> Please add a `setup-vlsi.sh` that sources the hammer sourceme and such.

Wait is this actually the right thing to do? The setup scripts are meant to be run only once to setup the environment (`build-toolchains.sh`, `firesim-setup.sh`). Its up to the user to source the correct bash script when they try to do something (`env.sh`, or `sourceme-f1-manager.sh`).

**Comment by colinschmidt** [comment]
Where do we should `vlsi-env.sh` then?

**Comment by jerryz123** [comment]
Do you mean like the `sourceme.sh` in eagle-chip? This seems like a project/user-specific env script that should be committed into the downstream tapeout-specific repos.

**Comment by colinschmidt** [comment]
Ok we can start without one but I think users will get annoyed.

**Comment by jwright6323** [comment]
Future annoyed user checking in

**Comment by jerryz123** [comment]
Testchipip test failing due to know branch issue, has since been resolved on dev.

`vlsi/sourceme.sh` will live in downstream tech-template.
 
On merge I will update the downstream tech-template, and work on merging the downstream tech-template with Beagle.

**Comment by colinschmidt** [comment]
I think we should just merge this sooner rather than later and then make more PRs to fix issues.
The ones I know are outstanding are in the makefile but I'd rather just make them new PRs.

**Pull request merged**